exemptions.py

```diff
@@ -0,0 +1,138 @@
+
+    Examples:
+
+    >>> from io import StringIO
+    >>> parse_conf_file(StringIO('# comment'))
+    {}
+
+    >>> parse_conf_file(StringIO('test_foo foo-id 2050-01-01 in prod never allow foo\\n'))
+    {'test_foo': {'foo-id': (0, '2050-01-01', 'in prod never allow foo')}}
+
+    >>> parse_conf_file(StringIO(
+    ... 'test_foo foo-id 2050-01-01 in prod never allow foo for foo-id\\n'
+    ... 'test_foo bar-id 2050-01-01 in prod never allow foo for bar-id'
+    ... )) == {
+    ... 'test_foo': {
+    ... 'foo-id': (0, '2050-01-01', 'in prod never allow foo for foo-id'),
+    ... 'bar-id': (1, '2050-01-01', 'in prod never allow foo for bar-id')}}
+    True
+
+
+    Short lines are skipped with a warning:
+
+    >>> parse_conf_file(StringIO('invalid line'))  # doctest:+ELLIPSIS
+    {}
+    >>> # UserWarning: Line 0: Skipping line with fewer than 4 whitespace delimited parts.
+
+    Invalid expirations dates are skipped with a warning:
+
+    >>> parse_conf_file(StringIO('test_foo foo-id 2050-01-AA in prod never allow foo'))
+    {}
+    >>> # UserWarning: Line 0: Skipping line with invalid expiration day '2050-01-AA'
+    >>> parse_conf_file(StringIO('test_foo foo-id 2000-01-01 in prod never allow foo'))
+    {}
+    >>> # UserWarning: Line 0: Skipping line with expiration day in the past '2000-01-01'
+
+    Duplicate test name and IDs are ignored with a warning:
+
+    >>> parse_conf_file(StringIO(
+    ... 'test_foo foo-id 2050-01-01 in prod never allow foo for foo-id\\n'
+    ... 'test_foo foo-id 2051-01-01 in prod never allow foo for bar-id'
+    ... ))
+    {'test_foo': {'foo-id': (0, '2050-01-01', 'in prod never allow foo for foo-id')}}
+    >>> # UserWarning: Line 1: Skipping line with duplicate test name and ID 'test_foo' 'foo-id'
+
+    Does not check that test name and IDs exist (since names might not
+    be collected and IDs can require an HTTP call).
+    """
+    # dict of test name to severity level
+    rules = {}
+
+    if not conf_fd:
+        return rules
+
+    for line_number, line in enumerate(conf_fd):
+        if line.startswith('#'):
+            continue
+
+        line_parts = line.split(maxsplit=3)
+        if len(line_parts) < 4:
+            warnings.warn('Line {}: Skipping line with fewer than 4 whitespace delimited parts.'.format(line_number))
+            continue
+
+        test_name, test_id, expiration, reason = line_parts[0], line_parts[1], line_parts[2], line_parts[3].strip('\n')
+        try:
+            expiration_date = datetime.strptime(expiration, '%Y-%m-%d')
+        except ValueError:
+            warnings.warn('Line {}: Skipping line with invalid expiration day {!r}'.format(line_number, expiration))
+            continue
+        if expiration_date < datetime.now():
+            warnings.warn('Line {}: Skipping line with expiration day in the past {!r}'.format(line_number, expiration))
+            continue
+
+        if test_name not in rules:
+            rules[test_name] = {}
```

Comment:
NB: Want to use a `defaultdict` for `rules`?

Reply:
Yeah, good idea. Updated.